pkg/volume/util/operationexecutor/operation_generator.go

```diff
@@ -1547,6 +1559,81 @@ func useCSIPlugin(vpm *volume.VolumePluginMgr, spec *volume.Spec) bool {
 	return false
 }
 
+func nodeUsingCSIPlugin(og *operationGenerator, spec *volume.Spec, nodeName types.NodeName) (bool, error) {
+	if len(nodeName) == 0 {
+		return false, goerrors.New("nodeName is empty")
+	}
+
+	migratable, err := og.volumePluginMgr.IsPluginMigratableBySpec(spec)
+	if err != nil {
+		return false, err
+	}
+	if !utilfeature.DefaultFeatureGate.Enabled(features.CSIMigration) ||
+		!utilfeature.DefaultFeatureGate.Enabled(features.CSINodeInfo) ||
+		!migratable {
+		return false, nil
+	}
+
+	kubeClient := og.volumePluginMgr.Host.GetKubeClient()
+	if kubeClient == nil {
+		// Assume standalone kubelet mode. In which case the we assume there is no CSINode
+		// to check and defer to useCSIPlugin for picking the backend
+		return true, nil
+	}
+
+	adcHost, ok := og.volumePluginMgr.Host.(volume.AttachDetachVolumeHost)
+	if !ok {
+		// This function is running not on the AttachDetachController
+		// We assume that Kubelet is servicing this function and therefore is
```

Comment:
is this saying that if the Host doesn't implement `CSINodeLister()` we assume we're running on the kubelet? that seems like a leap, for a few reasons:
* the `AttachDetachVolumeHost` interface could easily be satisfied by things that are not the attachdetachcontroller. if we want positive identification you need a method that is attachdetach-specific
* not implementing `AttachDetachVolumeHost` doesn't mean we're a kubelet, it just means we're not the attachdetach host.

can you clarify what the intended checks/responses are?

1. if we don't have a node name, return false?
2. if the volume can't be migrated, return false?
3. if the feature is disabled, return false?
4. if we're on the kubelet, return *true* to defer to `useCSIPlugin`?
5. if we're not on the kubelet, but have access to CSINodes, check that (and return false on lookup errors)?
6. if we're not on the kubelet, and don't have access to CSINodes, ... do what? I'd have expected to return false to be safe, but this code appears to be returning true in this case

Reply:
Ah good point. It always slips my mind that go interfaces are not strict when it comes to who is implementing what interfaces.

The function's intention is to run (and be useful on) *ONLY* the AttachDetachController to check whether the Kubelet has the volume plugin migrated on that particular Node for that particular volume.

intended checks and reasoning:
1) If we don't have Node Name -> return an error as this is unexpected
2) If the volume can't be migrated -> return false because it is assumed Kubelet volume also can't be migrated (if this is not true it's undefined behavior we do not support Kubelet version > master version)
3) If the feature is disabled -> return false because it is assumed Kubelet volume also should have feature disabled  (if this is not true it's undefined behavior we do not support Kubelet version > master version or this skew)
4) If we're on Kubelet [Unclear whether `kubeClient== nil` is the perfect check for that] -> return true because we do not need to "reach from ADC to Kubelet" and defer to `useCSIPlugin` which will check feature flags/translation lib etc.
5) If we're on ADC and have access to CSINodes -> Check the annotation, return error on API lookup (will retry in outer loop), return false if plugin is not found in annotation
6)  If we're not on the Kubelet and don't have access to CSINodes -> Should be returning an error

I think the most suspect area (which is why I commented it heavily) is the intended behavior here:
```
kubeClient := og.volumePluginMgr.Host.GetKubeClient()
if kubeClient == nil {
	// Assume standalone kubelet mode. In which case the we assume there is no CSINode
	// to check and defer to useCSIPlugin for picking the backend
	return true, nil
}

adcHost, ok := og.volumePluginMgr.Host.(volume.AttachDetachVolumeHost)
if !ok {
	// This function is running not on the AttachDetachController
	// We assume that Kubelet is servicing this function and therefore is
	// defer to useCSIPlugin for picking the backend
	return true, nil
}

csiNode, err := adcHost.CSINodeLister().Get(string(nodeName))
if err != nil {
	return false, err
}
```

What is meant to be captured is:
* If in standalone Kubelet mode, then `nodeUsingCSIPlugin == useCSIPlugin` since we are on the node (except we don't want to use CSINode to check) so we just return true and defer to `useCSIPlugin`.
* *The sketchy one*: If for some reason we're not running on the AttachDetachController (determined by type assertion) we again defer to `useCSIPlugin` for the same reason.
  * I guess the sketchy assumption here is that implicitly we know this function will only be called during Attach/Detach which can only be serviced by either ADC (in normal case) or Kubelet (if `enableControllerAttachDetach` is enabled)
* If all of the above is not true (we know we are running in Attach or Detach on some entity with access to CSINode) and yet we cannot `GET` the `CSINode` we care about, this is considered an error and will need to be retried in an outer loop

Reply:
Would `_, isKubelet := og.volumePluginMgr.Host.(volume.KubeletVolumeHost)` be a more definitive check of whether we were on the kubelet?

> If for some reason we're not running on the AttachDetachController (determined by type assertion) we again defer to use CSIPlugin for the same reason.

It seems like if we're running anywhere other than the kubelet, we need to have access to check CSINodes. The only place other than the kubelet happens to be the attachdetachcontroller today, but would it be more robust to require CSINode for all non-kubelet callers?

```go
if _, isKubelet := og.volumePluginMgr.Host.(volume.KubeletVolumeHost); isKubelet {
	// We assume that Kubelet is servicing this function and therefore
	// defer to useCSIPlugin for picking the backend
	return true, nil
}

csiNodeLister, ok := og.volumePluginMgr.Host.(volume.CSINodeListerHost)
if !ok {
	return false, fmt.Errorf("need access to CSINode objects to determine if the node is migrated to CSI")
}

csiNode, err := csiNodeLister.CSINodeLister().Get(string(nodeName))
if err != nil {
	return false, err
}
...
```

Reply:
also, if we *are* on the kubelet, does `useCSIPlugin` include checks of whether the driver is installed/available (since returning early skips those checks in nodeUsingCSIPlugin?

Reply:
> It seems like if we're running anywhere other than the kubelet, we need to have access to check CSINodes.

Don't think this is true actually. It's more of an ADC specific thing. Other storage components that run outside of Kubelet (pv controller, expand controller) will not need to make this check (as it is just for syncing attach/detach with mount/unmount). Therefore, I think this check needs to be done exclusively on the Attach Detach Controller, not *not* on the Kubelet.

> also, if we are on the kubelet, does useCSIPlugin include checks of whether the driver is installed/available (since returning early skips those checks in nodeUsingCSIPlugin)?

This is a good point. Kubelet will not check whether the driver is installed, only the whether the feature flags are flipped. The only benefit of checking for driver installation on the Kubelet is to be able to throw a better error message if the driver is not installed like this one: `return true, fmt.Errorf("in-tree plugin %s is migrated on node %s but driver %s is not installed", pluginName, string(nodeName), driverName)`.

Right now without the check the error will just be the generic CSI Driver not installed mount error.

However, it is questionable to me whether it's worth adding all the infrastructure for a Kubelet CSINode informer and extra logic right now in order to facilitate that. I could add a TODO to follow up on that? LMK if you think it is necessary to get in right now. @msau42 for opinions as well

Reply:
I think the existing "csi driver not found" error during mount is sufficient.

Reply:
> Other storage components that run outside of Kubelet (pv controller, expand controller) will not need to make this check (as it is just for syncing attach/detach with mount/unmount). Therefore, I think this check needs to be done exclusively on the Attach Detach Controller, not not on the Kubelet.

ok, then I think the current code is reasonable, but would recommend making the AttachDetachVolumeHost interface stronger (like add a marker `IsAttachDetachController()` method or something)

Reply:
done. Wish there was a better way to do strict interface assertions but this is an ok workaround. Thanks for digging deeper into the logic here :+1: